tools/wptrunner/wptrunner/wptrunner.py

```diff
@@ -235,143 +374,78 @@ def run_tests(config, test_paths, product, **kwargs):
-                    else:
-                        run_tests = test_loader.tests
-
-                    recording.pause()
-                    with ManagerGroup("web-platform-tests",
-                                      kwargs["processes"],
-                                      test_source_cls,
-                                      test_source_kwargs,
-                                      browser_cls,
-                                      browser_kwargs,
-                                      executor_cls,
-                                      executor_kwargs,
-                                      kwargs["rerun"],
-                                      kwargs["pause_after_test"],
-                                      kwargs["pause_on_unexpected"],
-                                      kwargs["restart_on_unexpected"],
-                                      kwargs["debug_info"],
-                                      not kwargs["no_capture_stdio"],
-                                      recording=recording) as manager_group:
-                        try:
-                            manager_group.run(test_type, run_tests)
-                        except KeyboardInterrupt:
-                            logger.critical("Main thread got signal")
-                            manager_group.stop()
-                            raise
-                        test_count += manager_group.test_count()
-                        unexpected_count += manager_group.unexpected_count()
-                        unexpected_pass_count += manager_group.unexpected_pass_count()
                 recording.set(["after-end"])
-                test_total += test_count
-                unexpected_total += unexpected_count
-                unexpected_pass_total += unexpected_pass_count
-                logger.info("Got %i unexpected results, with %i unexpected passes" %
-                            (unexpected_count, unexpected_pass_count))
+                logger.info(
+                    "Got %i unexpected results, with %i unexpected passes" %
+                    (counts["unexpected"], counts["unexpected_pass"]))
                 logger.suite_end()
-                if repeat_until_unexpected and unexpected_total > 0:
-                    break
-                if repeat_count == 1 and len(test_loader.test_ids) == skipped_tests:
-                    break
 
-    if test_total == 0:
-        if skipped_tests > 0:
-            logger.warning("All requested tests were skipped")
-        else:
-            if kwargs["default_exclude"]:
-                logger.info("No tests ran")
-                return True
-            else:
-                logger.critical("No tests ran")
-                return False
+                # determine the longest test suite runtime seen
+                longest_iteration_time = max(
+                    longest_iteration_time,
+                    datetime.now() - iteration_start)
```

Comment:
Vague musing; I wonder if we should 'pad' the iteration time by e.g. +10%, to avoid the case where we estimate we will complete but actually narrowly avoid it.

Reply:
I do think this is a good idea and not terribly hard to implement - I've added the change.